nautobot/docs/development/best-practices.md

```diff
@@ -74,7 +68,44 @@ The intended audience for the `validated_save()` convenience method is Job autho
 
 During execution, should model validation fail, `validated_save()` will raise `django.core.exceptions.ValidationError` in the normal Django fashion.
 
-## Slug Field
+### Field Naming in Data Models
+
+Model field names **must** always follow the following conventions:
+
+- Use lowercase letters, numbers, and underscores only
+- Separate words with underscores for readability/clarity
+- For foreign keys and their corresponding reverse-relations, match the `verbose_name` or `verbose_name_plural` of the related model
+
+Instead of:
+
+```python
+Rack.group
+DeviceType.consoleserverporttemplates
+Device.ipaddress_set
+```
+
+Use:
+
+```python
+Rack.rack_group
+DeviceType.console_server_port_templates
+Device.ip_addresses
+```
+
+#### Foreign Key `related_name` for Abstract Model Classes
+
++/- 2.0.0
+
+If an abstract model class defines a foreign key to a concrete model class, Django's default `related_name` functionality doesn't provide great options - the best you could normally do for a `related_name` on a `ForeignKey` from an abstract base class, for example, would be `"%(class)ss"` (potentially resulting in related names like `"devices"` or, less optimally, `"ipaddresss"`) or `"%(app_label)s_%(class)s_related"` (resulting in related names like `"dcim_device_related"` or `"ipam_ipaddress_related"`, which while at least consistent, are rather clunky).
```

Comment:
```suggestion
If an abstract model class defines a foreign key to a concrete model class, Django's default `related_name` functionality doesn't provide great options - the best you could normally do for a `related_name` on a `ForeignKey` from an abstract base class, for example, would be `"%(class)s"` (potentially resulting in related names like `"devices"` or, less optimally, `"ipaddresss"`) or `"%(app_label)s_%(class)s_related"` (resulting in related names like `"dcim_device_related"` or `"ipam_ipaddress_related"`, which while at least consistent, are rather clunky).
```

Reply:
```suggestion
If an abstract model class defines a foreign key to a concrete model class, Django's default `related_name` functionality doesn't provide great options - the best you could normally do for a `related_name` on a `ForeignKey` from an abstract base class, for example, would be `"%(class)ss"` (potentially resulting in related names like `"devices"` or, less optimally, `"ipaddresss"`) or `"%(app_label)s_%(class)s_related"` (resulting in related names like `"dcim_devices_related"` or `"ipam_ipaddresss_related"`, which while at least consistent, are rather clunky).
```

Reply:
This is correct as is, this feature is just confusing. :-) `%(class)ss` in the first example is actually `%(class)s` followed by a literal `s`, hence why you would get `device`+ `s` or `ipaddress` + `s`. Conversely, in the second example we're just using `%(app_label)s` and `%(class)s` (no extra `s`) so the singular names are correct. (Both of these are patterns that have been used in the code base before, they aren't totally random examples.)

Reply:
@glennmatthews So what you're saying is `%(class)s` => `Device` and `%(class)ss` => `Devices` where `class` is Device?

Reply:
Yeah this is old-school string parameterization/formatting, where the first `s` is denoting the type of the parameter `class` which is `str` in this case. If `class` were an int you would use `d`.

Reply:
> @glennmatthews So what you're saying is `%(class)s` => `Device` and `%(class)ss` => `Devices` where `class` is Device?

To be pedantic, `%(class)s` => `device` and `%(class)ss` => `devices`, yes.

Similarly for "IPAddress" this would result in `ipaddress` and `ipaddresss`.